language/libra-vm/src/errors.rs

```diff
@@ -0,0 +1,140 @@
+
+/// Converts particular Move abort codes to specific validation error codes for the prologue
+/// Any non-abort non-execution code is considered an invariant violation, specifically
+/// `UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION`
+pub fn convert_normal_prologue_error(status: VMStatus) -> VMStatus {
+    match status {
+        // TODO look at location? or remove this function entirely
+        VMStatus::MoveAbort(location, code) => {
+            let new_major_status = match code {
+                EACCOUNT_FROZEN => StatusCode::SENDING_ACCOUNT_FROZEN,
+                // Invalid authentication key
+                EBAD_ACCOUNT_AUTHENTICATION_KEY => StatusCode::INVALID_AUTH_KEY,
+                // Sequence number too old
+                ESEQUENCE_NUMBER_TOO_OLD => StatusCode::SEQUENCE_NUMBER_TOO_OLD,
+                // Sequence number too new
+                ESEQUENCE_NUMBER_TOO_NEW => StatusCode::SEQUENCE_NUMBER_TOO_NEW,
+                // Sequence number too new
+                EACCOUNT_DOES_NOT_EXIST => StatusCode::SENDING_ACCOUNT_DOES_NOT_EXIST,
+                // Can't pay for transaction gas deposit/fee
+                ECANT_PAY_GAS_DEPOSIT => StatusCode::INSUFFICIENT_BALANCE_FOR_TRANSACTION_FEE,
+                ETRANSACTION_EXPIRED => StatusCode::TRANSACTION_EXPIRED,
+                EBAD_CHAIN_ID => StatusCode::BAD_CHAIN_ID,
+                code => {
+                    crit!(
+                        "[libra_vm] Unexpected prologue move abort: {:?}::{:?}",
+                        location,
+                        code
+                    );
+                    return VMStatus::Error(StatusCode::UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION);
+                }
+            };
+            VMStatus::Error(new_major_status)
+        }
+        status @ VMStatus::ExecutionFailure { .. } | status @ VMStatus::Executed => status,
+        VMStatus::Error(code) => {
+            crit!("[libra_vm] Unexpected prologue error: {:?}", code);
+            VMStatus::Error(StatusCode::UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION)
+        }
+    }
+}
+
+/// Checks for only move aborts, execution failures, and successful execution.
+/// Any other errors are mapped to the invariant violation
+/// `UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION`
+pub fn convert_normal_epilogue_error(status: VMStatus) -> VMStatus {
+    match status {
+        status @ VMStatus::MoveAbort(_, _)
+        | status @ VMStatus::ExecutionFailure { .. }
+        | status @ VMStatus::Executed => status,
+        VMStatus::Error(code) => {
+            crit!("[libra_vm] Unexpected epilogue error: {:?}", code);
+            VMStatus::Error(StatusCode::UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION)
+        }
+    }
+}
+
+/// Converts move aborts and execution failures to `REJECTED_WRITE_SET`
+/// Any other errors are mapped to the invariant violation
+/// `UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION`
+pub fn convert_write_set_prologue_error(status: VMStatus) -> VMStatus {
```

Comment:
@runtian-zhou will know for sure, but I think the only expected results here are the following Move aborts:

- `INVALID_WRITESET_SENDER`
- `EPROLOGUE_SEQUENCE_NUMBER_TOO_OLD`
- `E_WSPROLOGUE_SEQUENCE_NUMBER_TOO_NEW` (question for Runtian: do we need both of these)?
- `EPROLOGUE_INVALID_ACCOUNT_AUTH_KEY`

Everything else should be an invariant violation.